File: c++/sample/virtual_call/virtual_call.cpp
#include <etsm.h>
#include <iostream>

using namespace etsm;

class Foo;

class FooState : public State<Foo>
{
public:
	FooState(Method enter, Method exit, Method tick = nullptr)
		: State<Foo>(enter, exit),
		tick(tick)
	{}

	void Tick(Foo* owner)
	{
		if (tick != nullptr)
			(owner->*tick)();
	}

private:
	Method tick;
};

class Foo
{
public:
	Foo()
		: sm(this),
		  a(nullptr, nullptr, &Foo::TickA),
		  b(nullptr, nullptr, &Foo::TickB)
	{}

	void Tick()
	{
		if (sm.GetCurrent())
			sm.GetCurrent()->Tick(this);
	}

	void Run()
	{
		Tick();
		sm.Transition(&a);
		Tick();
		sm.Transition(&b);
		Tick();
		sm.Transition(nullptr);
		std::cout << output << std::endl;
	}

private:
	StateMachine<Foo, FooState> sm;
	FooState a;
	FooState b;
	std::string output;

	void TickA() { output += " A "; }
	void TickB() { output += " B "; }
};

// Output: " A  B "
int main()
{
	Foo foo;
	foo.Run();
}

File: node/sample/virtual_call/virtual_call.js
var etsm = require('etsm');

class FooState extends etsm.State {
    constructor(enter, exit, tick) {
        super(enter, exit);
        this.Tick = tick;
    }
}

class Foo {
    constructor() {
        this.sm = new etsm.StateMachine(this);
        this.a = new FooState(null, null, this.TickA);
        this.b = new FooState(null, null, this.TickB);
    }

    TickA(){
        process.stdout.write(' A ');
    }

    TickB(){
        process.stdout.write(' B ');
    }

    Tick(){
        if ( this.sm.current != null && this.sm.current.Tick != null)
            this.sm.current.Tick.apply(this);
    }

    Test() {
        this.Tick();

        this.sm.Transition(this.a);
        this.Tick();

        this.sm.Transition(this.b);
        this.Tick();

        this.sm.Transition(null);
        this.Tick();
    }    
}

foo = new Foo();
foo.Test();